### PR TITLE
task(settings, content): Create configurable rollout rates for v2 key stretching

### PR DESCRIPTION
**Pull request by dschom**
## Because:
- We are prepping for a slow key stretching v2 rollout

## This Commit:
- Adds feature flag to to toggle on / off key stretching V2
- Adds experiment to enroll users into key stretching V2
   - Allows fxa-settings to check experiment state
- Adds query parameters to override feature flags
- Adds query parameters to force experiment
- Adds experiment for key stretch v2
- Exposes config to control experiment's roll out rate
- Set config defaults such that v2 key stretching is disabled

## Issue that this pull request solves

Closes: FXA-9080

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Other information (Optional)

A way to validate this is to change the config settings as desired. Run a `pm2 restart content`, and watch the network tab. V2 key stretching payloads will be delivered if the feature flag is flipped and the user gets enrolled in the experiment. Other than this, the v2 key stretch should go unnoticed by an end user.

### CODE REVIEW

File: packages/fxa-content-server/server/lib/configuration.js
Comment (pdehaan) [inline packages/fxa-content-server/server/lib/configuration.js:801]: Q: Is the rollout rate 0..1 or 0..100?

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts
Comment (pdehaan) [inline packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts:9]:
```suggestion
describe('Key Stretch Experiment Model', function () {
```

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts
Comment (pdehaan) [inline packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts:59]: Q: Why the double `JSON.stringify()`?

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts
Comment (pdehaan) [inline packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts:76]: Does this need a bug number?

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts
Comment (pdehaan) [inline packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts:87]: If something did throw an error, do we need to log it to Sentry or anything here?

File: packages/fxa-content-server/server/lib/configuration.js
Comment (dschom) [reply packages/fxa-content-server/server/lib/configuration.js:801]: 0 to 1

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts
Comment (dschom) [reply packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts:59]: That's just how it's stored. Doesn't work with a single stringify...

File: packages/fxa-content-server/server/lib/configuration.js
Comment (dschom) [reply packages/fxa-content-server/server/lib/configuration.js:801]: I'll update the doc so this is clear.

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts
Comment (dschom) [reply packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts:59]: Actually, let me double check this... This does seem strange. I know when testing I had to do this, but perhaps something else was off.

Comment (dschom) [reply packages/fxa-settings/src/models/experiments/key-stretch-experiment.test.ts:59]: Okay, manually verified this again. This aligns with the how the value is being stored by content server.

File: packages/fxa-content-server/app/scripts/lib/experiments/grouping-rules/index.js
Comment (LZoog) [inline packages/fxa-content-server/app/scripts/lib/experiments/grouping-rules/index.js:30]: Something seems funky here, it's been a while since I've set one of these up but we've had several experiments in the past with configurable rollout rates so I wouldn't think we'd need this... Did you look at [the docs](https://mozilla.github.io/ecosystem-platform/reference/experiments-ab-testing) we have or past experiments we've removed? 🤔 Maybe we do need this, just double checking.

**Review by LZoog — COMMENTED** [review]
This mostly LGTM with only one real question, I'll circle back soon and do some local testing. Code changes look like the new feature flag takes precedence over experiment status, except if `stretch=2` is set in the URL for test purposes, and users are only considered "in v2" if the feature flag is on and they are enrolled in the experiment. 👍

I think unfortunately you might need to make some tweaks for `SigninUnblock` once #16376 has landed...

Thanks for all the in-code comments!

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts
Comment (pdehaan) [reply packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts:28]: Don't know if there are any risks&hellip; but if this _is_ just for dev/test purposes, should we add any checks to make sure we aren't running this code path on production env (unless that is overly complex due to Docker-ing)?

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts
Comment (dschom) [reply packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts:67]: So my reasoning was that this part felt like a bit of a hack anyway... We are essentially going behind the experiment module codes back and accessing state it is setting. Also notice the kinda bizarre double JSON parse / stringify calls.

In my mind, FXA-9183 (the follow for porting experiments module) is where we should support experimentation would address proper access to local storage. There were already places, like metrics that was doing it this way, and it seemed quickest / most direct.

File: packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts
Comment (dschom) [reply packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts:28]: Yep, it's for dev/test only. I gennerally avoid conditional logic against the env as this is much better accomplished with config.

I personally don't see this being a risk since there are other query parameters that can turn this on. One thing to keep in mind, is that we might want to turn it on for a single step, but not the entire flow. That's another reason to have this override.

I can see an argument for putting this if statement after the feature flag check though... so I'll do that since it caught some scrutiny.

Comment (LZoog) [reply packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts:28]: Yeah, my thinking was we would want to be able to disable it entirely with the feature flag (SRE being able to do this without a retag is super nice), but in that case I don't think we even need the param since we could use the force experiment params. 

I'm unsure of test cases we could benefit from by letting it override the feature flag, but also not sure what real security concerns there could be (what's the worst case scenario... some small sector of users might try messing with it and might get their account in a weird state?) so I'll leave it up to you to decide.

Comment (dschom) [reply packages/fxa-settings/src/models/experiments/key-stretch-experiment.ts:28]: I don't think there is much risk. I think the real use case is that we want to test a single step of the login flow with quick stretch enabled, but not the rest. If we use the experiment parameters, then we end up enrolling and having to go through content server. This query parameter flag was definitely useful for testing isolated behavior when V2 key stretching is turned on. 

I will definitely give the feature flag precedence though!

File: packages/fxa-settings/src/pages/Signup/container.tsx
Comment (dschom) [reply packages/fxa-settings/src/pages/Signup/container.tsx:213]: Good catch on the missing dep! 

For the nit, I sorta like passing the config, because the resulting code feels more obvious to (ie I know exactly where the feature flag comes form...) So I might ignore, unless you would really rather have the other way.

File: packages/fxa-content-server/app/scripts/lib/experiments/grouping-rules/index.js
Comment (dschom) [reply packages/fxa-content-server/app/scripts/lib/experiments/grouping-rules/index.js:30]: I looked at all the current experiments and docs and didn't see anything. They had hard coded rollout rates. I didn't look at previous experiments, but I did perused the experiments module and didn't find another way to set this. This actually a descent part of the effort, cause I sort of assumed there was a way to do this.

**Review by LZoog — APPROVED** [review]

**Pull request merged**
